src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,277 @@
+		podSpec := &sts.Spec.Template.Spec
+		if err := r.modifyPodImage(podSpec, newImage); err != nil {
+			return err
+		}
+
+		if err := r.partitionUpdateImage(ctx, sts, newImage); err != nil {
+			return err
+		}
+
+		// Update is complete for all pods (and all are ready). Set upgrading status to false.
+		if err := r.updateUpgradeStatus(ctx, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) updateUpgradeStatus(
+	ctx context.Context, upgrading bool,
+) error {
+	if !reflect.DeepEqual(upgrading, r.pandaCluster.Status.Upgrading) {
+		r.pandaCluster.Status.Upgrading = upgrading
+		if err := r.Status().Update(ctx, r.pandaCluster); err != nil {
+			r.logger.Error(err, "Failed to update RedpandaClusterStatus")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) partitionUpdateImage(
+	ctx context.Context, sts *appsv1.StatefulSet, newImage string,
+) error {
+	r.logger.Info("Start statefulset image update")
+
+	// When a StatefulSet's partition number is set to `i`, only Pods with ordinal
+	// greater than or equal to `i` will be updated.
+	// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#partitions
+	for ordinal := *sts.Spec.Replicas - 1; ordinal >= 0; ordinal-- {
+		// Update() on statefulset has not been called yet in this run, however,
+		// this could be a retry call in which case we skip the current partition.
+		poderr := r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr == nil {
+			r.logger.Info("Pod already updated, skip", "ordinal", ordinal)
+			continue
+		}
+
+		// Continue only if error is due to pod not ready, or unchanged image.
+		if !errors.Is(poderr, errContainerHasWrongImage) && !errors.Is(poderr, errPodNotReady) {
+			return poderr
+		}
+
+		// Ideally, at this point we'd wait until all Pods are in ready state.
+		// However, if something went wrong with a previous image change and the pod
+		// is in bad state, e.g., due to wrong or erroneous image, the pod will
+		// keep restarting and require manual intervention. Instead, for now we
+		// skip the readiness requirement and "force" the image update.
+
+		if err := r.rollingUpdatePartition(ctx, sts, ordinal); err != nil {
+			return err
+		}
+
+		poderr = r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr != nil {
+			msg := fmt.Sprintf("Pod not ready, will retry - ordinal %d", ordinal)
+			r.logger.Info(msg)
+
+			return &NeedToReconcileError{RequeueAfter: requeueDuration, Msg: msg}
+		}
+
+		// Give time to redpanda pod to rejoin cluster.
+		// TODO: In a revision, an API call to redpanda is made to help us decide when to continue.
+		// If the answer is negative, we would requeue as above.
+		r.logger.Info("Sleeping for 1 minute before updating the next pod, or completion")
+		time.Sleep(time.Minute)
+
+		// After Update(), refresh the statefulset for the next iteration (or other use).
+		err := r.Get(ctx, r.Key(), sts)
```

Comment:
Maybe use `r.LastObservedState` instead of sts?

Reply:
I use `sts` because otherwise I'd just add `sts := r.LastObservedState` in each function, so I preferred to make it obvious that we operate on the sts by passing it around.